drivers/dma/dma_sedi.c

```diff
@@ -0,0 +1,533 @@
+	}
+
+	return ret;
+
+ERR:
+	LOG_ERR("dma transfer failed");
+	return ret;
+}
+
+static int dma_sedi_stop(const struct device *dev, uint32_t channel)
+{
+	const struct dma_sedi_config_info *const info = DEV_CFG(dev);
+	struct dma_sedi_driver_data *const data = DEV_DATA(dev);
+	struct dma_config *config = &(data->dma_configs[channel]);
+
+	LOG_DBG("stopping dma: %p, %d", dev, channel);
+	sedi_dma_abort_transfer(info->peripheral_id, channel);
+
+	free_blk_nodes(config);
+	return 0;
+}
+
+static const struct dma_driver_api dma_funcs = { .config = dma_sedi_chan_config,
+						 .start = dma_sedi_start,
+						 .stop = dma_sedi_stop,
+						 .reload = dma_sedi_reload,
+						 .get_status = NULL
+};
+
+static int dma_sedi_init(const struct device *dev)
+{
+	const struct dma_sedi_config_info *const info = DEV_CFG(dev);
+	sedi_dma_t dma_dev = info->peripheral_id;
+	uint32_t irq_flag = 0;
+
+	switch (dma_dev) {
+#if DT_NODE_HAS_STATUS(DT_DRV_INST(0), okay)
+	case SEDI_DMA_0:
+#if DT_INST_IRQ_HAS_CELL(0, sense)
+		irq_flag = DT_INST_IRQ(0, sense);
+#endif
+		IRQ_CONNECT(DT_INST_IRQN(0),
+			    DT_INST_IRQ(0, priority),
+			    dma_isr, (void *)SEDI_DMA_0, irq_flag);
+		irq_enable(DT_INST_IRQN(0));
+		break;
+#endif
+#if DT_NODE_HAS_STATUS(DT_DRV_INST(1), okay)
+	case SEDI_DMA_1:
+#if DT_INST_IRQ_HAS_CELL(1, sense)
+		irq_flag = DT_INST_IRQ(1, sense);
+#endif
+		IRQ_CONNECT(DT_INST_IRQN(1),
+			    DT_INST_IRQ(1, priority),
+			    dma_isr, (void *)SEDI_DMA_1, irq_flag);
+		irq_enable(DT_INST_IRQN(1));
+		break;
+#endif
+#if DT_NODE_HAS_STATUS(DT_DRV_INST(2), okay)
+	case SEDI_DMA_2:
+#if DT_INST_IRQ_HAS_CELL(2, sense)
+		irq_flag = DT_INST_IRQ(2, sense);
+#endif
+		IRQ_CONNECT(DT_INST_IRQN(2),
+			    DT_INST_IRQ(2, priority),
+			    dma_isr, (void *)SEDI_DMA_2, irq_flag);
+		irq_enable(DT_INST_IRQN(2));
+		break;
+#endif
+	default:
+		return -EIO;
+	}
+#if defined(CONFIG_PM_DEVICE)
+	struct dma_sedi_driver_data *const data = DEV_DATA(dev);
+
+	data->power_status = PM_DEVICE_STATE_ACTIVE;
+#endif
+	return 0;
+}
+
+#ifdef CONFIG_PM_DEVICE
+
+static bool is_dma_busy(sedi_dma_t dev)
+{
+	sedi_dma_status_t chn_status;
+
+	for (int chn = 0; chn < DMA_CHANNEL_NUM; chn++) {
+		sedi_dma_get_status(dev, chn, &chn_status);
+		if (chn_status.busy == 1) {
+			return true;
+		}
+	}
+	return false;
+}
+
+static int dma_change_device_power(const struct device *dev,
+		enum pm_device_action action)
+{
+	struct dma_sedi_driver_data *const data = DEV_DATA(dev);
+	const struct dma_sedi_config_info *const info = DEV_CFG(dev);
+	sedi_dma_t dma_dev = info->peripheral_id;
+	int ret;
+
+	sedi_power_state_t state;
+
+	switch (action) {
+	case PM_DEVICE_ACTION_RESUME:
+		state = SEDI_POWER_FULL;
+		break;
+	case PM_DEVICE_ACTION_SUSPEND:
+		if (is_dma_busy(dma_dev)) {
+			return -EBUSY;
+		}
+		state = SEDI_POWER_SUSPEND;
+		break;
+
+	default:
+		return -ENOTSUP;
+	}
+
+	for (uint8_t chn = 0; chn < DMA_CHANNEL_NUM; chn++) {
+		ret = sedi_dma_set_power(dma_dev, chn, state);
```

Comment:
Can each channel really be power gated? That's neat if so, and might mean a little rethinking could make better use of this in terms of the power management API here. Mostly the entire DMA device is treated as a power managed device the way the API is setup today.

Reply:
HAL uses this to power gate per channel. Here, we just follow this.